Adm1266 add initial support

tachicialex opened the pull request:
Add initial pmbus probing driver for the adm1266 Cascadable
Super Sequencer with Margin Control and Fault Recording.
Driver is currently using the pmbus_core, creating sysfs
files under hwmon for inputs: vh1->vh4 and vp1->vp13.

nunojsa commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+	{ },
+};
+MODULE_DEVICE_TABLE(i2c, adm1266_id);
+
```
Any reason to use this table? Can this actually be probed rather than dt? I would just stick with `of_device_id` and ditch this...

commodo replied:
this has been adapted from the adm1275;
the current usecase for the adm1266 is data-centers; which implies more ACPI;
we may need to consider things to be a bit more neutral between OF & ACPI [for the ADM1266 at least];

nunojsa replied:
But for acpi don't we actually need `acpi_device_id`? As far as I see, `spi_device_id` is just being used to get `driver_data` and that can be done with `of_device_id`. But I may be missing something since I never used ACPI core...

Either way, I would not add things because of "future expectations". That leads to "dead" and "unneeded" code to stay around (if we actually never implement the piece of code)... So I would either add what needs to be added (that justifies `spi_device_id`) or remove it and add it later...

commodo replied:
there should be some ACPI code that bridges/handles `of_device_id` as well;

we'll see what the hwmon maintainer says though

nunojsa replied:
> there should be some ACPI code that bridges/handles of_device_id as well;

That seems to be the case. Look [here](https://elixir.bootlin.com/linux/latest/source/drivers/acpi/bus.c#L822). That is why that I don't really think we need the `spi_device_id` table...

nunojsa commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+	for (mid = adm1266_id; mid->name[0]; mid++) {
+		if (!strncasecmp(mid->name, block_buffer, strlen(mid->name)))
+			break;
+	}
```
nitpick: new line

nunojsa commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+
+	for (mid = adm1266_id; mid->name[0]; mid++) {
+		if (!strncasecmp(mid->name, block_buffer, strlen(mid->name)))
+			break;
```
What are you trying to accomplish with this code? Not sure if I'm missing something but you already have `const struct i2c_device_id *id` passed to you. Why not just comparing that to whatever you get from the device? Moreover, IMHO, i would just stick with of_device_id...

commodo replied:
this also comes from the adm1275 [which was written by the hwmon/pmbus maintainer];
i'm also a bit paranoid about these checks; especially for ADM1266 these can be different, because they depend a lot on the FW on the device;

i'm tempted to remove them on the first iteration of this driver;
not sure;

nunojsa replied:
I'm still not sure why we need to iterate also. We should already have `mid->name` for the device being probed that we could use directly to do the `strncasecmp()`...

nunojsa commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+		return -ENODEV;
+	}
+
+	wr_buffer[0] = 32;
```
Just curious. Why do we need this before calling `adm1266_block_wr`. Any change to have this inside the function?

tachicialex replied:
Here 32 is not the write packet length, it specifies how many bytes of data i want from the device (some SmBus Block WR commands need additional parameters like data length/indexes).

nunojsa replied:
Can't we make that a parameter of the function? It looks a bit weird that we need to this and then call the function... But in the end is up to you :+1:

commodo commented on drivers/hwmon/pmbus/Kconfig:
```diff
@@ -26,6 +26,16 @@ config SENSORS_PMBUS
 	  be called pmbus.
 
+config SENSORS_ADM1266
+	tristate "Analog Devices ADM1266"
```
maybe make the name `Analog Devices ADM1266 Sequencer`

commodo replied:
any follow-up on this?

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+
+#include "pmbus.h"
+
+enum chips { adm1266 };
```
it's a good idea to make enums:
```
enum chips {
      adm1266,
};
```

when adding new parts you just add new entries

commodo commented on Documentation/hwmon/adm1266:
```diff
@@ -0,0 +1,34 @@
+Kernel driver adm1266
+=====================
```
filename should be `Documentation/hwmon/adm1266.rst`

nunojsa replied:
You should also make sure the doc builds... If you didn't do that, take a look at Documention/doc-guide

commodo commented on Documentation/hwmon/adm1266:
```diff
@@ -0,0 +1,34 @@
+
+Supported chips:
+  * Analog Devices ADM1266
+    Datasheet: https://www.analog.com/media/en/technical-documentation/data-sheets/ADM1266.pdf
```
any idea if we need to add a line `Prefix: 'adm1266` here?
something like
```
Supported chips:
  * Analog Devices ADM1266
    Prefix: 'adm1266'
    Datasheet: https://www.analog.com/media/en/technical-documentation/data-sheets/ADM1266.pdf
```

commodo commented on Documentation/hwmon/adm1266:
```diff
@@ -0,0 +1,34 @@
+integrated 12 bit SAR ADC, accessed using a PMBus interface.
+
+The driver is a client driver to the core PMBus driver. Please see
+Documentation/hwmon/pmbus for details on PMBus client drivers.
```
document should be `Documentation/hwmon/pmbus.rst`
note that you should compare against a more recent linux tree, not the ADI master;
the ADI master seems a bit behind some doc-stuff;

commodo commented on drivers/hwmon/pmbus/Kconfig:
```diff
@@ -26,6 +26,16 @@ config SENSORS_PMBUS
+	default n
+	help
+	  If you say yes here you get hardware monitoring support for Analog
+	  Devices ADM1266.
```
`Devices ADM1266.` -> `Devices ADM1266 Cascadable Super Sequencer.`

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+};
+
+static int adm1266_block_wr(struct i2c_client *client, u8 cmd, u8 wr_len,
+			    u8 *data_w, u8 *data_r)
```
maybe also pass the size/len for `data_r`?

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+	if (ret < 0)
+		return ret;
+
+	memcpy(data_r, &smbus_data.block[1], smbus_data.block[0]);
```
here, you are not protecting memcpy() agaisnt invalid accesses;
`memcpy(data_r, &smbus_data.block[1], smbus_data.block[0]);`

` smbus_data.block[0]` could be whatever the HW says; and could cause the memcpy() to go outside valid memory

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+		return -ENOMEM;
+	data->id = mid->driver_data;
+
+	ret = adm1266_config(&data->info);
```
will you add more stuff to this `adm1266_config()`?

so far, it is returning 0, so, the error path seems unreachable

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+		return -ENODEV;
+
+	wr_buffer[0] = 32;
+	ret = adm1266_block_wr(client, PMBUS_MFR_ID, 1, wr_buffer,
```
why not use `i2c_smbus_read_block_data()` here directly?

tachicialex replied:
I would get an -EPROTO as MFR_ID is a Block WR command.

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,146 @@
+ * Copyright 2020 Analog Devices Inc.
+ */
+
+#include <linux/bitops.h>
```
curios: are all these includes required?

commodo commented on Documentation/devicetree/bindings/hwmon/adi,adm1266.yaml:
```diff
@@ -0,0 +1,47 @@
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Analog Devices ADM1266 Cascadable Super Sequencer with Margin
+Control and Fault Recording
```
is this correct for the yaml format?

tachicialex commented:
Changelog V1 -> V2:
- removed Kconfig default n
- added function comments for adm1266_block_wr
- now export gpios and pdios using the gpio provider api
- added debugfs for go_command and read_state
- removed multiple chip ids support, this driver will support only adm1266

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -6,24 +6,281 @@
+ * @data_r: Byte array into which data will be read; big enough to hold
+ *	the data returned by the slave.  SMBus allows at most 32 bytes.
+ *
+ * Different from Block Read as it sends data and
```
nitpick: you can probably make the comment wider/closer to the 80 col limit

commodo commented:
this PR doesn't build the driver;
maybe add it the `drivers/hwmon/Kconfig.adi`

commodo commented:
checkpatch also has some complaints;  
https://travis-ci.org/github/analogdevicesinc/linux/jobs/668573200#L281
```
ImportError: No module named git
WARNING: Missing or malformed SPDX-License-Identifier tag in line 1
#27: FILE: Documentation/hwmon/adm1266.rst:1:
+Kernel driver adm1266
```

i'll fix the `ImportError: No module named git` error;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -6,24 +6,281 @@
 static int adm1266_probe(struct i2c_client *client,
 			 const struct i2c_device_id *id)
 {
 	struct pmbus_driver_info *info;
```
won't the compiler complain here that this isn't used anymore?

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -6,24 +6,281 @@
+	}
+}
+
+static int adm1266_config_gpio(struct adm1266_data *data)
```
we probably need to do something like:
```
#if IS_ENABLED(CONFIG_GPIOLIB)
static int adm1266_config_gpio(struct adm1266_data *data)
{
......
}
#else
static inline int adm1266_config_gpio(struct adm1266_data *data)
{
          return 0;
}
#endif

commodo replied:
umm; maybe the `#if IS_ENABLED(CONFIG_GPIOLIB)` needs to be extended to all the new GPIO functions

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -260,6 +263,49 @@ static int adm1266_config_gpio(struct adm1266_data *data)
+{
+	struct dentry *root;
+
+	root = debugfs_create_dir("adm1266_debugfs", NULL);
```
maybe add a distinctive information for the folder name;
maybe the i2c-addr?

`root = debugfs_create_dir("adm1266_%d{i2c_addr}_debugfs", NULL);`

you can have multiple chips on the same system, so this would throw an error/warning on the 2nd chip that gets probed;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -6,24 +6,281 @@
+
+/* ADM1266 GPIO defines */
+#define ADM1266_GPIO_NR			9
+#define ADM1266_GPIO_FUNCTIONS(x)      FIELD_GET(BIT(0), x)
```
this should be indented with tabs to be consisten;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -6,24 +6,281 @@
+	unsigned int pmbus_cmd;
+	int ret;
+
+	if (offset < ADM1266_GPIO_NR)
```
you can probably group this offset check here like:
```
if (offset >= ADM1266_GPIO_NR) {
         pmbus_cmd = ADM1266_PDIO_STATUS;
         offset -= ADM1266_GPIO_NR;
} else {
         pmbus_cmd = ADM1266_GPIO_STATUS;
}
```

then you can do just:  `return test_bit(adm1266_gpio_mapping[offset][1], &pins_status);` at the end;

commodo replied:
wait; nevermind here;
this isn't correct;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -260,6 +263,49 @@ static int adm1266_config_gpio(struct adm1266_data *data)
+	int ret;
+
+	val = FIELD_GET(GENMASK(4, 0), val);
+	ret = i2c_smbus_write_word_data(data->client, ADM1266_GO_COMMAND, val);
```
maybe just do `return i2c_smbus_write_word_data(data->client, ADM1266_GO_COMMAND, val);`

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,62 @@
+
+static const struct i2c_device_id adm1266_id[] = {
+	{ "adm1266", 0 },
+	{ },
```
some upstream people now complain that null terminators should not have a trailing comma
it's a good idea to remove it now;

tachicialex commented:
Changelog v2 -> v3:
- moved adm1266_block_wr to pmbus_block_wr in pmbus_core
- debugfs_create_dir debugfs_create_dir adds i2c addr to folder name
- added patch to enable building
- added blackbox reading feature through nvmem api
- added #if IS_ENABLED(CONFIG_GPIOLIB) to all gpio related functions

commodo commented on drivers/hwmon/pmbus/Kconfig:
```diff
@@ -4,7 +4,7 @@
 menuconfig PMBUS
 	tristate "PMBus support"
-	depends on I2C
+	depends on I2C && CRC8
```
i wouldn't move add `CRC8` here;
ADM1266 depends on it, but I don't think other PMBUS devices do or should;
maybe, do `IS_ENABLED(CRC8)` somewhere in the ADM1266 driver; some users may want to use this chip without the CRC8 support;

tachicialex replied:
Doing pmbus_do_probe will automatically activate packet error checking  (if the device supports it), no matter the user's preference and pmbus_core uses i2c_smbus's hardcoded weird crc8 implementation. 
https://github.com/analogdevicesinc/linux/blob/18a7db3f76c09b1df9f4aa35fb3182ee387dc881/drivers/hwmon/pmbus/pmbus_core.c#L2013

commodo commented:
you also need to update the 2 defconfigs `arch/arm/configs/zynq_xcomm_adv7511_defconfig` and `adi_zynqmp_defconfig`
see: https://travis-ci.org/github/analogdevicesinc/linux/jobs/677732825#L3962

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -0,0 +1,62 @@
+
+static const struct of_device_id adm1266_of_match[] = {
+	{ .compatible = "adi,adm1266" },
+	{ },
```
remove comma for null terminator?

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -6,24 +6,242 @@
+
+/* ADM1266 GPIO defines */
+#define ADM1266_GPIO_NR			9
+#define ADM1266_GPIO_FUNCTIONS(x)	FIELD_GET(BIT(0), x)
```
generally: `FIELD_GET()` isn't really useful for single-bit definitions; it's more useful for multi-bit stuff;
so, if you do `#define ADM1266_GPIO_FUNCTIONS   BIT(0)`
then check `if (gpio_config & ADM1266_GPIO_FUNCTIONS)`

but i guess for this case you can leave it as-is for now;

commodo commented on drivers/hwmon/pmbus/adm1266.c:
```diff
@@ -6,24 +6,242 @@
+	struct i2c_client *client;
+};
+
+static const unsigned int adm1266_gpio_mapping[ADM1266_GPIO_NR][2] = {
```
should these gpio defs go under `#if IS_ENABLED(CONFIG_GPIOLIB)`?

commodo commented on drivers/hwmon/pmbus/pmbus_core.c:
```diff
@@ -57,6 +58,8 @@
 
 #define PMBUS_NAME_SIZE		24
 
+DECLARE_CRC8_TABLE(pmbus_crc_table);
```
this seems to declare the crc table for the whole driver;
shouldn't this be declared per-device?
in the idea that each device has it's own table?

tachicialex replied:
From the SmBus spec 3.0:
"The PEC may be calculated in any way that conforms
to a CRC-8 represented by the polynomial, C(x) = x8 + x2 + x1 + 1"

commodo commented on drivers/hwmon/pmbus/pmbus_core.c:
```diff
@@ -186,6 +189,98 @@ int pmbus_set_page(struct i2c_client *client, int page)
+
+	dma_buf = kzalloc(PMBUS_BLOCK_MAX + 3, GFP_KERNEL);
+	if (!dma_buf)
+		return;
```
maybe return -ENOMEM and treat this error?

commodo commented on drivers/hwmon/pmbus/pmbus_core.c:
```diff
@@ -186,6 +189,98 @@ int pmbus_set_page(struct i2c_client *client, int page)
+	u8 crc = 0;
+	int ret;
+
+	i2c_pmbus_try_get_dmabuf(&msgs[0]);
```
i'm confused;
`struct i2c_msg msgs[2]` has some buffers attached to it, from the stack, but then this comes and allocs some new buffers?

tachicialex replied:
This method is taken from the SmBus Linux implementation.

nunojsa replied:
Do we want to unconditional use dma safe buffer here? How bit are typically these messages?
And why not using i2c apiS `i2c_get_dma_safe_msg_buf()` and `i2c_put_dma_safe_msg_buf`?

tachicialex replied:
Most messages will be of 255 length bytes (when writing fw/configs or reading blackboxes).
ADM1266 data-sheet (pag. 55): "Read back user data length, maximum value of N = 252".

nunojsa replied:
Ok, then makes sense to me to always go with dma safe buffers. Still, Why not using the i2c functions?  If you stick with `i2c_pmbus_try_get_dmabuf`, I would return an error and make the use of dma safe buffer mandatory (meaning, just bail out if you cannot allocate a buffer)...
